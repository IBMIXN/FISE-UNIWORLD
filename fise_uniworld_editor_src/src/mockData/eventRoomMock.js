const mockEventRoom = {
  title: "mockTitle",
  eventDate: "2021-03-17T00:00:00.000Z",
  meetingTables: [],
  background: "mockBackground",
  scene: "mockScene",
};

const mockEventRoomUpdate = {
  title: "newMockTitle",
  eventDate: "2021-02-27T12:00:00.000Z",
  meetingTables: [],
  background: "newMockBackground",
  scene: "newMockScene",
};

const mockMeetingTable = {
  title: "mockTitle",
  type: "RoundMeetingTable",
  logoUrl: "https://via.placeholder.com/150",
  zoomUrl: "https://zoom.us/",
  posX: 0,
  posY: 0,
};

export { mockEventRoom, mockEventRoomUpdate, mockMeetingTable };
