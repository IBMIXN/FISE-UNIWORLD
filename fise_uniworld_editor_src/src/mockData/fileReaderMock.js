export default class FileReaderMock {
  onerror = jest.fn();
  onload = jest.fn();
  addEventListener = (_, fn) => fn({ target: { result: {} } });
  readAsDataURL = jest.fn();
}
