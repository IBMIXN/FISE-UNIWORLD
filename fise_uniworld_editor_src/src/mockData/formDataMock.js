export default class FormDataMock {
  append = jest.fn();
}
