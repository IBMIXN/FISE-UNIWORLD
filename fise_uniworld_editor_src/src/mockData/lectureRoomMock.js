const mockLectureRoom = {
  title: "mockTitle",
  module: "mockModule",
  lecturer: "mockLecturer",
  startTime: "2021-03-17T09:00:00.000Z",
  endTime: "2021-03-17T10:00:00.000Z",
  firstSlideUrl: "mockUrl",
  numSlides: 1,
};

const mockLectureRoomUpdate = {
  title: "newMockTitle",
  module: "newMockModule",
  lecturer: "newMockLecturer",
  startTime: "2021-02-27T14:00:00.000Z",
  endTime: "2021-02-27T15:00:00.000Z",
  firstSlideUrl: "newMockUrl",
  numSlides: 3,
};

export { mockLectureRoom, mockLectureRoomUpdate };
