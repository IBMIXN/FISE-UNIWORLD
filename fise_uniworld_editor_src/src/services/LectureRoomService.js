import { axiosConfig } from "../config";

class LectureRoomService {
  url = "/lecturerooms";

  async get() {
    return (await axiosConfig.get(this.url)).data;
  }

  async create(data) {
    return (await axiosConfig.post(this.url, data)).data;
  }

  async update(id, data) {
    return (await axiosConfig.put(`${this.url}/${id}`, data)).data;
  }

  async delete(id) {
    return await axiosConfig.delete(`${this.url}/${id}`);
  }
}

export default new LectureRoomService();
