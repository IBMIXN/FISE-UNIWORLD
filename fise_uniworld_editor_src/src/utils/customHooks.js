import React from "react";

const useReducerState = (defaultState) => {
  const reducer = (state, newState) => ({ ...state, ...newState });
  return React.useReducer(reducer, defaultState);
};

export { useReducerState };
