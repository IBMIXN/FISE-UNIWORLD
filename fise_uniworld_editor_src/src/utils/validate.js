import moment from "moment";

const validateEventRoom = (eventRoom) => {
  return (
    eventRoom.title &&
    eventRoom.scene &&
    eventRoom.background &&
    moment(eventRoom.eventDate).isValid() &&
    Array.isArray(eventRoom.meetingTables)
  );
};

const validateLectureRoom = (lectureRoom) => {
  return (
    lectureRoom.title &&
    lectureRoom.module &&
    lectureRoom.lecturer &&
    lectureRoom.firstSlideUrl &&
    Number.isInteger(lectureRoom.numSlides) &&
    moment(lectureRoom.startTime).isValid() &&
    moment(lectureRoom.endTime).isValid()
  );
};

export { validateEventRoom, validateLectureRoom };
