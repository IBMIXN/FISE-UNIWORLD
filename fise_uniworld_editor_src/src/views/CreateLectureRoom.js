import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import { Container, Typography, Snackbar } from "@material-ui/core";
import MuiAlert from "@material-ui/lab/Alert";
import LectureRoomService from "../services/LectureRoomService";
import { useReducerState } from "../utils/customHooks";
import { validateLectureRoom } from "../utils/validate";
import LectureRoomForm from "../components/LectureRoomForm";

const useStyles = makeStyles((theme) => ({
  mainSection: {
    paddingTop: theme.spacing(5),
  },
}));

const CreateLectureRoom = (props) => {
  const classes = useStyles();
  const [alert, setAlert] = useReducerState({
    open: false,
    message: "",
    type: "success",
  });

  const showAlert = (msg, type) => {
    setAlert({ open: true, message: msg, type: type });
  };

  const hideAlert = () => {
    setAlert({ open: false });
  };

  const handleSubmit = async (data) => {
    if (validateLectureRoom(data)) {
      try {
        const createdLectureRoom = await LectureRoomService.create(data);
        props.history.push({
          pathname: `/lecturerooms/${createdLectureRoom._id}`,
          state: { created: true },
        });
      } catch (err) {
        showAlert("Failed to create lecture room. Please try again later.", "error");
        console.log(err);
      }
    } else {
      showAlert("Please make sure you have entered all the fields", "warning");
    }
  };

  return (
    <Container className={classes.mainSection}>
      <Snackbar
        open={alert.open}
        autoHideDuration={5000}
        onClose={hideAlert}
        anchorOrigin={{ vertical: "top", horizontal: "center" }}
      >
        <MuiAlert elevation={6} variant="filled" onClose={hideAlert} severity={alert.type}>
          {alert.message}
        </MuiAlert>
      </Snackbar>
      <Typography variant="h4">Create Lecture Room</Typography>
      <LectureRoomForm onSubmit={handleSubmit} submitText="Create Lecture Room" />
    </Container>
  );
};

export default CreateLectureRoom;
