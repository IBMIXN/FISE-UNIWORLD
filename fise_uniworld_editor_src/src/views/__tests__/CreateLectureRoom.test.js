import React from "react";
import { shallow } from "enzyme";
import Snackbar from "@material-ui/core/Snackbar";
import CreateLectureRoom from "../CreateLectureRoom";
import LectureRoomService from "../../services/LectureRoomService";
import { mockLectureRoom } from "../../mockData/lectureRoomMock";
import { useReducerState } from "../../utils/customHooks";
import { validateLectureRoom } from "../../utils/validate";

jest.mock("../../services/LectureRoomService");
jest.mock("../../utils/validate");
jest.mock("../../utils/customHooks", () => {
  return {
    useReducerState: jest.fn(),
  };
});

describe("Create Lecture Room view test", () => {
  const history = { push: jest.fn() };

  const flushPromises = () => new Promise(setImmediate);
  const alert = {
    open: false,
    message: "",
    type: "success",
  };
  const setAlert = jest.fn();
  beforeEach(() => {
    useReducerState.mockImplementation(() => [alert, setAlert]);
    jest.spyOn(console, "log").mockImplementation(() => {});
  });

  it("should match snapshot", () => {
    const component = shallow(<CreateLectureRoom history={history} />);
    expect(component).toMatchSnapshot();
  });

  it("should correctly handle submit lecture room form event", async () => {
    validateLectureRoom.mockImplementationOnce(() => true);
    LectureRoomService.create.mockImplementationOnce(() =>
      Promise.resolve({ ...mockLectureRoom, _id: "mockId" })
    );
    const component = shallow(<CreateLectureRoom history={history} />);
    component.find("LectureRoomForm").first().prop("onSubmit")(mockLectureRoom);
    await flushPromises();
    expect(validateLectureRoom).toBeCalledWith(mockLectureRoom);
    expect(LectureRoomService.create).toBeCalledWith(mockLectureRoom);
    expect(history.push).toBeCalledWith({
      pathname: "/lecturerooms/mockId",
      state: { created: true },
    });
    expect(setAlert).not.toBeCalled();
  });

  it("should correctly handle submit lecture room form event when validation fails", async () => {
    validateLectureRoom.mockImplementationOnce(() => false);
    const component = shallow(<CreateLectureRoom history={history} />);
    component.find("LectureRoomForm").first().prop("onSubmit")(mockLectureRoom);
    await flushPromises();
    expect(validateLectureRoom).toBeCalledWith(mockLectureRoom);
    expect(setAlert).toBeCalledWith({
      open: true,
      message: "Please make sure you have entered all the fields",
      type: "warning",
    });
    expect(LectureRoomService.create).not.toBeCalled();
    expect(history.push).not.toBeCalled();
  });

  it("should correctly handle submit lecture room form event when an error is thrown", async () => {
    validateLectureRoom.mockImplementationOnce(() => true);
    LectureRoomService.create.mockImplementationOnce(() => Promise.reject());
    const component = shallow(<CreateLectureRoom history={history} />);
    component.find("LectureRoomForm").first().prop("onSubmit")(mockLectureRoom);
    await flushPromises();
    expect(validateLectureRoom).toBeCalledWith(mockLectureRoom);
    expect(LectureRoomService.create).toBeCalledWith(mockLectureRoom);
    expect(setAlert).toBeCalledWith({
      open: true,
      message: "Failed to create lecture room. Please try again later.",
      type: "error",
    });
    expect(history.push).not.toBeCalled();
  });

  it("should correctly handle close alert dialog", () => {
    const component = shallow(<CreateLectureRoom history={history} />);
    component.find(Snackbar).first().prop("onClose")();
    expect(setAlert).toBeCalledWith({ open: false });
  });
});
