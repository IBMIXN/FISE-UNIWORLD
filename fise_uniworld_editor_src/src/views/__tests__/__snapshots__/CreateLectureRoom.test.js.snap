// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Lecture Room view test should match snapshot 1`] = `
<WithStyles(ForwardRef(Container))
  className="makeStyles-mainSection-1"
>
  <WithStyles(ForwardRef(Snackbar))
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
    autoHideDuration={5000}
    onClose={[Function]}
    open={false}
  >
    <WithStyles(ForwardRef(Alert))
      elevation={6}
      onClose={[Function]}
      severity="success"
      variant="filled"
    />
  </WithStyles(ForwardRef(Snackbar))>
  <WithStyles(ForwardRef(Typography))
    variant="h4"
  >
    Create Lecture Room
  </WithStyles(ForwardRef(Typography))>
  <LectureRoomForm
    onSubmit={[Function]}
    submitText="Create Lecture Room"
  />
</WithStyles(ForwardRef(Container))>
`;
